ntp-udp/src/lib.rs

```diff
@@ -178,25 +212,50 @@ const fn control_message_space<T>() -> usize {
     (unsafe { libc::CMSG_SPACE((std::mem::size_of::<T>()) as _) }) as usize
 }
 
-fn recv(socket: &std::net::UdpSocket, buf: &mut [u8]) -> io::Result<(usize, Option<NtpTimestamp>)> {
+fn recv(
+    socket: &std::net::UdpSocket,
+    buf: &mut [u8],
+) -> io::Result<(usize, SocketAddr, Option<NtpTimestamp>)> {
     let mut buf_slice = IoSliceMut::new(buf);
 
-    // could be on the stack if const extern fn is stable
     let mut control_buf = [0; control_message_space::<[libc::timespec; 3]>()];
+    let mut addr = MaybeUninit::<libc::sockaddr_storage>::uninit();
     let mut mhdr = libc::msghdr {
         msg_control: control_buf.as_mut_ptr().cast::<libc::c_void>(),
         msg_controllen: control_buf.len(),
         msg_iov: (&mut buf_slice as *mut IoSliceMut).cast::<libc::iovec>(),
         msg_iovlen: 1,
         msg_flags: 0,
-        msg_name: std::ptr::null_mut(),
-        msg_namelen: 0,
+        msg_name: addr.as_mut_ptr().cast::<libc::c_void>(),
+        msg_namelen: size_of::<libc::sockaddr_storage>() as u32,
     };
 
     // loops for when we receive an interrupt during the recv
     let flags = 0;
     let bytes_read = receive_message(socket, &mut mhdr, flags)? as usize;
 
+    let addr = unsafe { addr.assume_init() };
+    let sock_addr = match addr.ss_family as i32 {
+        libc::AF_INET => {
+            // kernel assures us this conversion is safe
+            let sin = &addr as *const _ as *const libc::c_void as *const libc::sockaddr_in;
+            let sin = unsafe { &*sin };
+
+            // no direct (u32, u16) conversion is available, so we convert the address first
+            let addr = Ipv4Addr::from(sin.sin_addr.s_addr);
+            SocketAddr::from((addr, sin.sin_port))
+        }
+        libc::AF_INET6 => {
+            // kernel assures us this conversion is safe
+            let sin = &addr as *const _ as *const libc::c_void as *const libc::sockaddr_in6;
+            let sin = unsafe { &*sin };
+            SocketAddr::from((sin.sin6_addr.s6_addr, sin.sin6_port))
+        }
+        _ => {
+            unreachable!("We never constructed a non-ip socket");
+        }
+    };
```

Comment:
would be useful as a helper function for the ifaddr stuff, but can do that in that PR too

Reply:
Let's defer this to a separate PR (or the sendtimestamp PR)